content/blog/2024-01-10-xilem-2024.md

```diff
@@ -0,0 +1,88 @@
+
+As mentioned in a [previous blog](https://raphlinus.github.io/rust/gui/2022/07/15/next-dozen-guis.html), any given application will have a set of requirements. Is Web one of the main deployment targets? Will there be embedded video? Is there a need to integrate with some other subsystems, such as a game engine? Each of these has profound implications. Some problem spaces (compilers are an example) are “smooth,” in that continual refinement will lead to fairly similar outcomes no matter the starting point, but over time I’ve come to the conclusion that UI is especially lumpy. I believe this contributes to the continuing pattern of new UI toolkits coming out every couple months or so; the author surveys what’s available, finds none that match the specific set of requirements, and creates a new one.
+
+In some ways, Xilem is in the same pattern, but there are *some* things we’re doing differently. For one, it’s explicitly a research project, with knowledge about how to build UI in Rust an outcome possibly as important as the codebase. We’re trying to be systematic about finding the best ways to do things, which ideally will make some solutions more general. And, increasingly, we’re designing things as modular layers that can be swapped out. With luck, even if people don’t end up adopting the entire stack, there will be some useful components for the ecosystem.
+
+A good overview of the opportunities and challenges in Rust UI, particularly for making mobile apps, is [parallel futures in mobile application development](https://wingolog.org/archives/2023/06/15/parallel-futures-in-mobile-application-development) by Andy Wingo. The linked blog posts in that series also go into considerable detail about JavaScript and TypeScript based approaches, which have important lessons as well.
+
+### Declarative UI patterns
+
+The trend in UI programming has been overwhelmingly away from a soup of mutable objects with interlinked references and toward various declarative or reactive patterns. That’s a good thing for Rust UI, because the former is very un-ergonomic in Rust.
+
+In the JavaScript world, there is a Cambrian explosion of reactive techniques. React is by far the most popular, but there is a strong sentiment it’s possible to be better, particularly on performance, but also robustness, as the logic for deciding [when to re-render](https://www.joshwcomeau.com/react/why-react-re-renders/) is fickle, and can [miss re-renders](https://arxiv.org/abs/2310.04368). Elm deserves special mention, as its avoidance of global shared mutability makes it particularly easy to adapt to Rust, so it serves as the reactive core of the [iced], [relm], and [vizia] UI toolkits, as well as yew on the WASM side. But there is also Ember Octane (based on Adapton), a number of fine-grained reactivity approaches such as SolidJS, and many, many more (not to mention the incredible diversity of solutions for state management).
```

Comment:
The later reference to yew is linked, but either the first, or both, should be linked. I looked it up before I realized it was linked below. Also, it may make sense to link Ember Octane, and maybe SolidJS. Just to be consistent.